### PR TITLE
feat: implement local repositories for listings and reviews

### LINKED ISSUE #192: Add new local repository classes for reviews and listings

### PR DESCRIPTION
**Pull request by rami-mallah**
> Disclaimer: This PR has 800 lines affected, with only a little less than 200 lines of new implementation. The rest is 450 lines of tests and 150 lines of documentation.

## Summary

This PR implements local repository classes that provide offline access to reviews and rental listings using Room database. These repositories implement the same interfaces as their Firestore equivalent, enabling seamless switching between local and remote data sources in future PRs. This closes #192.

### Added Files

- `ReviewsRepositoryLocal`: Room-backed implementation of `ReviewsRepository`
  - Provides offline access to reviews stored in local Room database
  - Implements all repository methods including vote operations

- `RentalListingRepositoryLocal`: Room-backed implementation of `RentalListingRepository`
  - Provides offline access to rental listings stored in local Room database
  - Implements location-based filtering using in-memory distance calculations

### Implementation details

1. Offline Creation Prevention: Both repositories throw `UnsupportedOperationException` when `getNewUid()` is called, preventing users from creating new content offline. However, `addReview()` and `addRentalListing()` still work for syncing existing content with UIDs from remote sources.

2. Interface Consistency: Both local repositories implement the same interfaces as their Firestore equivalent, ensuring they can be used interchangeably.

3. Location Filtering: `getAllRentalListingsByLocation()` performs in-memory filtering after fetching all listings, matching the behavior of the Firestore implementation.

## Tests

All tests use Room's in-memory database and follow the same patterns as existing DAO tests. Tests verify:
- Successful operations
- Error handling
- Edge cases (empty lists, not found, etc.)
- All branches in vote logic

## Next Steps

This PR sets the foundation for another PR, which will:
- Create hybrid repositories that combine local and remote data sources
- Update repository providers to switch between local and remote based on connectivity
- Implement data syncing from Firestore to Room

## Notes
All documentation was generated with the help of AI. 
This PR message was written with the help of AI.
Some tests were generated by AI based on the specified test cases provided to it.

### CODE REVIEW

File: app/src/main/java/com/android/mySwissDorm/model/rental/RentalListingRepositoryLocal.kt [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+  /**
+   * Retrieves a specific rental listing by its unique identifier.
+   *
+   * @param rentalPostId The unique identifier of the rental listing.
+   * @return The rental listing with the specified identifier.
+   * @throws Exception if the rental listing is not found.
+   */
+  override suspend fun getRentalListing(rentalPostId: String): RentalListing {
+    return rentalListingDao.getRentalListing(rentalPostId)?.toRentalListing()
+        ?: throw Exception("RentalListingRepositoryLocal: Rental listing $rentalPostId not found")
```
Comment (sophieur444) [inline app/src/main/java/com/android/mySwissDorm/model/rental/RentalListingRepositoryLocal.kt]: returning “not found” by throwing a plain Exception is unclear and forces callers to treat it like any other error. In plain terms: callers can’t easily tell if a lookup failed because the listing is missing or because something else went wrong. Better options are to return a nullable value (so callers can check for null) or throw a specific, named exception (for ex, RentalListingNotFoundException) so the absence case can be handled cleanly.

**Review by sophieur444 — CHANGES_REQUESTED** [review]
Well done, you added Room-backed local repositories for Reviews and RentalListing with full CRUD, vote toggling (up/down/remove) and self-vote protection, distance-based filtering with error logging, and comprehensive Robolectric tests covering happy paths and edge cases.

Some suggestions to consider (not necessarily mandatory):
Throw more specific exceptions (for example NoSuchElementException or a custom NotFoundException).
Use fixed timestamps or a test clock instead of Timestamp.now() in tests.
Move geospatial filtering into SQL/Room or add indexes for better performance on large datasets.
Wrap multi-step vote updates in transactions to ensure atomicity.

You can implement them or not, they are not critical. Well done with the coding! 
this review was done with the help of ai

File: app/src/main/java/com/android/mySwissDorm/model/rental/RentalListingRepositoryLocal.kt (hunk 1 above)
Comment (rami-mallah) [reply app/src/main/java/com/android/mySwissDorm/model/rental/RentalListingRepositoryLocal.kt]: That's true! I'll change those generic `Exception`s to more meaningful ones like `NoSuchElementException` and `IllegalArgumentException`.

File: app/src/test/java/com/android/mySwissDorm/model/review/ReviewsRepositoryLocalTest.kt [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+      ownerId: String = "user-1",
+      title: String = "Test Review",
+      residencyName: String = "Vortex",
+      upvotedBy: Set<String> = emptySet(),
+      downvotedBy: Set<String> = emptySet()
+  ): Review {
+    return Review(
+        uid = uid,
+        ownerId = ownerId,
+        postedAt = Timestamp.now(),
```
[the remark this answers is not in the document]
Comment (rami-mallah) [reply app/src/test/java/com/android/mySwissDorm/model/review/ReviewsRepositoryLocalTest.kt]: Changed that too
